compiler/noirc_frontend/src/parser/parser.rs

```diff
@@ -91,1764 +44,504 @@ use traits::where_clause;
+    current_token_span: Span,
+    previous_token_span: Span,
 }
 
-fn check_statements_require_semicolon(
-    statements: Vec<(Statement, (Option<Token>, Span))>,
-    _span: Span,
-    emit: &mut dyn FnMut(ParserError),
-) -> Vec<Statement> {
-    let last = statements.len().saturating_sub(1);
-    let iter = statements.into_iter().enumerate();
-    vecmap(iter, |(i, (statement, (semicolon, span)))| {
-        statement.add_semicolon(semicolon, span, i == last, emit)
-    })
-}
-
-/// Parse an optional ': type'
-fn optional_type_annotation<'a>() -> impl NoirParser<UnresolvedType> + 'a {
-    ignore_then_commit(just(Token::Colon), parse_type()).or_not().map_with_span(|r#type, span| {
-        r#type.unwrap_or(UnresolvedTypeData::Unspecified.with_span(span))
-    })
-}
-
-fn module_declaration() -> impl NoirParser<TopLevelStatementKind> {
-    attributes().then(item_visibility()).then_ignore(keyword(Keyword::Mod)).then(ident()).validate(
-        |((attributes, visibility), ident), span, emit| {
-            let attributes = validate_secondary_attributes(attributes, span, emit);
-            TopLevelStatementKind::Module(ModuleDeclaration {
-                visibility,
-                ident,
-                outer_attributes: attributes,
-            })
-        },
-    )
-}
+impl<'a> Parser<'a> {
+    pub fn for_lexer(lexer: Lexer<'a>) -> Self {
+        Self::new(TokenStream::Lexer(lexer))
+    }
 
-fn use_statement() -> impl NoirParser<TopLevelStatementKind> {
-    item_visibility()
-        .then_ignore(keyword(Keyword::Use))
-        .then(use_tree())
-        .map(|(visibility, use_tree)| TopLevelStatementKind::Import(use_tree, visibility))
-}
+    pub fn for_tokens(mut tokens: Tokens) -> Self {
+        tokens.0.reverse();
+        Self::new(TokenStream::Tokens(tokens))
+    }
 
-fn rename() -> impl NoirParser<Option<Ident>> {
-    ignore_then_commit(keyword(Keyword::As), ident()).or_not()
-}
+    pub fn for_str(str: &'a str) -> Self {
+        Self::for_lexer(Lexer::new(str))
+    }
 
-fn use_tree() -> impl NoirParser<UseTree> {
-    recursive(|use_tree| {
-        let simple = path::path_no_turbofish().then(rename()).map(|(mut prefix, alias)| {
-            let ident = prefix.pop().ident;
-            UseTree { prefix, kind: UseTreeKind::Path(ident, alias) }
-        });
-
-        let list = {
-            let prefix = maybe_empty_path().then_ignore(just(Token::DoubleColon));
-            let tree = use_tree
-                .separated_by(just(Token::Comma))
-                .allow_trailing()
-                .delimited_by(just(Token::LeftBrace), just(Token::RightBrace))
-                .map(UseTreeKind::List);
-
-            prefix.then(tree).map(|(prefix, kind)| UseTree { prefix, kind })
+    fn new(tokens: TokenStream<'a>) -> Self {
+        let mut parser = Self {
+            errors: Vec::new(),
+            tokens,
+            token: SpannedToken::default(),
+            next_token: SpannedToken::default(),
```

Comment:
If this is the only case we needed SpannedToken::default and Token::default for, I'd rather we just manually created the filler tokens here to prevent the rest of the codebase from using default on these types.

Reply:
Agreed, removed :-)